File: builtin/python/baihu/env.py
import os
import json
import urllib.request
import urllib.error

def _get_headers():
    token = os.environ.get("BHPKG_OPENAPI_TOKEN") or os.environ.get("OPENAPI_TOKEN") or os.environ.get("BHPKG_NOTIFY_TOKEN")
    if not token:
        raise RuntimeError("没有正确配置或缺少 BHPKG_OPENAPI_TOKEN 环境变量以使用 env 函数。请在白虎面板的任务设置中配置这些 Key。")
    return {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {token}"
    }

def _get_envs_url():
    url = os.environ.get("BHPKG_OPENAPI_URL") or os.environ.get("OPENAPI_URL")
    if url:
        return url
    
    notify_url = os.environ.get("BHPKG_NOTIFY_URL", "http://localhost:8052/api/v1/notify/send")
    for target in ["/api/v1/notify/send/", "/api/v1/notify/send", "/api/v1/notify/", "/api/v1/notify"]:
        if target in notify_url:
            return notify_url.replace(target, "/open2api/v1/env")
            
    return "http://localhost:8052/open2api/v1/env"

def _request(url, method="GET", data=None):
    headers = _get_headers()
    payload = None
    if data is not None:
        payload = json.dumps(data).encode("utf-8")
    
    req = urllib.request.Request(url, data=payload, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req) as resp:
            body = resp.read().decode("utf-8")
            if not body:
                return {}
            parsed = json.loads(body)
            if isinstance(parsed, dict) and parsed.get("code") is not None and parsed.get("code") != 200:
                msg = parsed.get("msg") or parsed.get("message") or "未知错误"
                raise RuntimeError(f"请求失败 [{parsed.get('code')}]: {msg}")
            return parsed
    except urllib.error.HTTPError as e:
        err_body = e.read().decode("utf-8")
        try:
            err_json = json.loads(err_body)
            msg = err_json.get("msg") or err_json.get("message") or err_body
        except Exception:
            msg = err_body
        raise RuntimeError(f"请求失败 [{e.code}]: {msg}")
    except Exception as e:
        raise RuntimeError(f"请求发生异常: {e}")

def get_envs():
    """
    获取所有的环境变量列表。
    """
    url = f"{_get_envs_url()}/all"
    res = _request(url, "GET")
    return res.get("data", [])

def get_env(name):
    """
    根据变量名获取环境变量。如果不存在则返回 None。
    """
    envs = get_envs()
    for env in envs:
        if env.get("name") == name:
            return env
    return None

def add_envs(envs_list):
    """
    批量添加环境变量。
    envs_list: 包含环境变量字典的列表，如 [{"name": "KEY", "value": "VAL", "remark": "备注"}]
    """
    url = _get_envs_url()
    added_envs = []
    for env in envs_list:
        if "name" not in env or "value" not in env:
            raise ValueError("环境变量必须包含 'name' 和 'value'")
        res = _request(url, "POST", env)
        if "data" in res:
            added_envs.append(res["data"])
    return added_envs

def add_env(name, value, remark="", type="normal", hidden=True, enabled=True):
    """
    添加单个环境变量。
    """
    url = _get_envs_url()
    payload = {
        "name": name,
        "value": value,
        "remark": remark,
        "type": type,
        "hidden": hidden,
        "enabled": enabled
    }
    res = _request(url, "POST", payload)
    return res.get("data")

def update_env(id, name, value, remark=None, type=None, hidden=None, enabled=None):
    """
    根据 ID 更新环境变量。
    """
    url = f"{_get_envs_url()}/{id}"
    payload = {}
    if name is not None: payload["name"] = name
    if value is not None: payload["value"] = value
    if remark is not None: payload["remark"] = remark
    if type is not None: payload["type"] = type
    if hidden is not None: payload["hidden"] = hidden
    if enabled is not None: payload["enabled"] = enabled

    res = _request(url, "PUT", payload)
    return res.get("data")

def delete_envs(ids):
    """
    批量删除环境变量。
    """
    for fid in ids:
        delete_env(fid)

def delete_env(id):
    """
    根据 ID 删除指定的环境变量。
    """
    url = f"{_get_envs_url()}/{id}"
    _request(url, "DELETE")
    return True

File: builtin/nodejs/env.js
const http = require('http');
const https = require('https');
const { URL } = require('url');

/**
 * 内部 API 请求辅助函数
 */
function request(urlStr, method = 'GET', data = null) {
    const token = process.env.BHPKG_OPENAPI_TOKEN || process.env.OPENAPI_TOKEN || process.env.BHPKG_NOTIFY_TOKEN;
    if (!token) {
        throw new Error(`没有正确配置或缺少 BHPKG_OPENAPI_TOKEN 环境变量以使用 env 函数。请在白虎面板的任务设置中配置这些 Key。`);
    }

    const parsedUrl = new URL(urlStr);
    const protocol = parsedUrl.protocol === 'https:' ? https : http;
    
    let payload = '';
    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`
    };

    if (data !== null) {
        payload = JSON.stringify(data);
        headers['Content-Length'] = Buffer.byteLength(payload);
    }

    const options = {
        hostname: parsedUrl.hostname,
        port: parsedUrl.port,
        path: parsedUrl.pathname + (parsedUrl.search || ''),
        method: method,
        headers: headers
    };

    return new Promise((resolve, reject) => {
        const req = protocol.request(options, (res) => {
            let body = '';
            res.setEncoding('utf8');
            res.on('data', (chunk) => body += chunk);
            res.on('end', () => {
                if (res.statusCode >= 200 && res.statusCode < 300) {
                    try {
                        const parsed = body ? JSON.parse(body) : {};
                        if (parsed && typeof parsed === 'object' && parsed.code !== undefined && parsed.code !== 200) {
                            reject(new Error(`请求失败 [${parsed.code}]: ${parsed.msg || parsed.message || '未知错误'}`));
                        } else {
                            resolve(parsed);
                        }
                    } catch (e) {
                        resolve(body);
                    }
                } else {
                    let errMsg = body;
                    try {
                        const parsed = JSON.parse(body);
                        errMsg = parsed.msg || parsed.message || body;
                    } catch(e) {}
                    reject(new Error(`请求失败 [${res.statusCode}]: ${errMsg}`));
                }
            });
        });
        
        req.on('error', (e) => reject(e));
        if (payload) {
            req.write(payload);
        }
        req.end();
    });
}

function getEnvsUrl() {
    const url = process.env.BHPKG_OPENAPI_URL || process.env.OPENAPI_URL;
    if (url) return url;

    const notifyUrl = process.env.BHPKG_NOTIFY_URL || 'http://localhost:8052/api/v1/notify/send';
    const targets = ['/api/v1/notify/send/', '/api/v1/notify/send', '/api/v1/notify/', '/api/v1/notify'];
    for (const target of targets) {
        if (notifyUrl.includes(target)) {
            return notifyUrl.replace(target, '/open2api/v1/env');
        }
    }
    return 'http://localhost:8052/open2api/v1/env';
}

/**
 * 获取所有的环境变量列表
 */
async function getEnvs() {
    const url = `${getEnvsUrl()}/all`;
    const res = await request(url, 'GET');
    return res.data || [];
}

/**
 * 根据变量名获取环境变量，不存在则返回 null
 */
async function getEnv(name) {
    const envs = await getEnvs();
    for (const env of envs) {
        if (env.name === name) {
            return env;
        }
    }
    return null;
}

/**
 * 批量添加环境变量
 */
async function addEnvs(envsList) {
    const url = getEnvsUrl();
    const addedEnvs = [];
    for (const env of envsList) {
        if (!env.name || !env.value) {
            throw new Error("环境变量必须包含 'name' 和 'value'");
        }
        const res = await request(url, 'POST', env);
        if (res.data) {
            addedEnvs.push(res.data);
        }
    }
    return addedEnvs;
}

/**
 * 添加单个环境变量
 */
async function addEnv(name, value, remark = "", type = "normal", hidden = true, enabled = true) {
    const url = getEnvsUrl();
    const payload = {
        name,
        value,
        remark,
        type,
        hidden,
        enabled
    };
    const res = await request(url, 'POST', payload);
    return res.data;
}

/**
 * 根据 ID 更新环境变量
 */
async function updateEnv(id, name, value, remark = null, type = null, hidden = null, enabled = null) {
    const url = `${getEnvsUrl()}/${id}`;
    const payload = {};
    if (name !== null) payload.name = name;
    if (value !== null) payload.value = value;
    if (remark !== null) payload.remark = remark;
    if (type !== null) payload.type = type;
    if (hidden !== null) payload.hidden = hidden;
    if (enabled !== null) payload.enabled = enabled;

    const res = await request(url, 'PUT', payload);
    return res.data;
}

/**
 * 批量删除环境变量
 */
async function deleteEnvs(ids) {
    for (const id of ids) {
        await deleteEnv(id);
    }
    return true;
}

/**
 * 根据 ID 删除指定环境变量
 */
async function deleteEnv(id) {
    const url = `${getEnvsUrl()}/${id}`;
    await request(url, 'DELETE');
    return true;
}

module.exports = {
    getEnvs,
    getEnv,
    addEnvs,
    addEnv,
    updateEnv,
    deleteEnvs,
    deleteEnv
};
